### PR TITLE
Random: use random_bytes() on PHP 7

### PR DESCRIPTION
**Pull request by JanTvrdik**
Not sure whether the generateBytes() refactoring is worth it. It's still a bit ugly.

### CODE REVIEW

**Comment by JanTvrdik** [comment]
Also: both `random_bytes` and `mcrypt_create_iv` emit warnings if `$length` is zero.

File: src/Utils/Random.php [hunk 1]
```diff
@@ -52,10 +43,41 @@ public static function generate($length = 10, $charlist = '0-9a-z')
...
+			$bytes = openssl_random_pseudo_bytes($length);
+		} elseif ($strategy === 2 && function_exists('mcrypt_create_iv')) {
+			$bytes = mcrypt_create_iv($length, MCRYPT_DEV_URANDOM);
+		} elseif ($strategy === 3 && @is_readable('/dev/urandom')) {
+			$bytes = file_get_contents('/dev/urandom', FALSE, NULL, -1, $length);
+		} elseif ($strategy === 4) {
+			static $cache;
+			$bytes = str_pad('', $length, $cache ?: $cache = md5(serialize($_SERVER), TRUE));
+		} elseif ($strategy === 5) {
+			throw new Nette\InvalidStateException();
```
Comment (fprochazka) [inline src/Utils/Random.php]: you might wanna add here some explanatory message

File: src/Utils/Random.php [hunk 2]
```diff
@@ -52,10 +43,41 @@ public static function generate($length = 10, $charlist = '0-9a-z')
...
+		} elseif ($strategy === 1 && function_exists('openssl_random_pseudo_bytes')) {
+			$bytes = openssl_random_pseudo_bytes($length);
+		} elseif ($strategy === 2 && function_exists('mcrypt_create_iv')) {
+			$bytes = mcrypt_create_iv($length, MCRYPT_DEV_URANDOM);
+		} elseif ($strategy === 3 && @is_readable('/dev/urandom')) {
+			$bytes = file_get_contents('/dev/urandom', FALSE, NULL, -1, $length);
+		} elseif ($strategy === 4) {
+			static $cache;
+			$bytes = str_pad('', $length, $cache ?: $cache = md5(serialize($_SERVER), TRUE));
+		} elseif ($strategy === 5) {
```
Comment (fprochazka) [inline src/Utils/Random.php]: this should probably by >= 5

Comment (enumag) [inline src/Utils/Random.php]: I think it should be `} else {`.

Comment (dg) [inline src/Utils/Random.php]: This is really magic implementation

**Comment by hrach** [comment]
I would use while cycle, no need for recursion.

File: src/Utils/Random.php (hunk 2 above)
Comment (JanTvrdik) [inline src/Utils/Random.php]: That's why it is private. I originally wanted to make the method public.

**Comment by JanTvrdik** [comment]
I hate both the original and my own implementation :disappointed:

**Comment by dg** [comment]
Whats wrong with current implementation? It's straightforwardhttps://github.com/nette/utils/blob/master/src/Utils/Random.php#L34-L45.

**Comment by JanTvrdik** [comment]
> Whats wrong with current implementation? 

You evaluate `empty($rand3)` more times than it is necessary. Unfortunately every alternative implementation I have came up with either had similar problem or was too complicated.

**Comment by JanTvrdik** [comment]
We currently have inconsistency between `php.ini` for Windows and Linux. Should this be fixed as well?

**Comment by dg** [comment]
Of course, but are they really inconsistent?

**Comment by JanTvrdik** [comment]
> but are they really inconsistent?

That's what I don't know. Are some extensions on Linux available even with `-n`? 

* `php-unix.ini` enables JSON
* `php-win.ini` enabled mbstring, gd2, fileinfo and openssl

**Comment by dg** [comment]
Are tests failing?

**Comment by JanTvrdik** [comment]
No, but in `Random` case if you don't have openssl extension, you will just not be able to test the `openssl_random_pseudo_bytes` strategy. You would need to generate code coverage on CI to see what has actually been covered.

**Comment by dg** [comment]
Random is used to work in all environments, so it is really hard to test it completely.

Btw I don't see openssl here https://github.com/nette/utils/blob/master/tests/php-win.ini

**Comment by JanTvrdik** [comment]
@dg Yes, I added openssl in the first commit.

> it is really hard to test it completely

Yes, but with the improved tests (see the 1st commit) it tests first 4 strategies on Travis. If you run them on Windows, it will test the last strategy as well.

**Comment by milo** [comment]
> Are some extensions on Linux available even with -n? 

These are compile args from Debian rules files for PHP 5.6.7 cli. All others extensions must be installed as a `.so` library.
```
COMMON_CONFIG=--build=$(DEB_BUILD_GNU_TYPE) \
                --host=$(DEB_HOST_GNU_TYPE) \
                --sysconfdir=/etc \
                --localstatedir=/var \
                --mandir=/usr/share/man \
                --disable-debug \
                --with-regex=php \
                --disable-rpath \
                --disable-static \
                --with-pic \
                --with-layout=GNU \
                --with-pear=/usr/share/php \
                --enable-calendar \
                --enable-sysvsem \
                --enable-sysvshm \
                --enable-sysvmsg \
                --enable-bcmath \
                --with-bz2 \
                --enable-ctype \
                --with-db4 \
                --with-qdbm=/usr \
                --without-gdbm \
                --with-iconv \
                --enable-exif \
                --enable-ftp \
                --with-gettext \
                --enable-mbstring \
                --with-onig=/usr \
                --with-pcre-regex=/usr \
                --enable-shmop \
                --enable-sockets \
                --enable-wddx \
                --with-libxml-dir=/usr \
                --with-zlib \
                --with-kerberos=/usr \
                --with-openssl=/usr \
                --enable-soap \
                --enable-zip \
                --with-mhash=yes \
                --with-system-tzdata \
                --with-mysql-sock=/var/run/mysqld/mysqld.sock \
                $(CONFIGURE_DTRACE_ARGS)

CFLAGS="$(CFLAGS)" PROG_SENDMAIL="$(PROG_SENDMAIL)" ../configure \
                --prefix=/usr --disable-cgi \
                --with-config-file-path=/etc/php5/cli \
                --with-config-file-scan-dir=/etc/php5/cli/conf.d \
                $(COMMON_CONFIG) \
                --without-mm \
                --disable-pdo \
                --without-mysql --without-sybase-ct \
                --without-mssql --without-sqlite3 --enable-pcntl \
                --with-libedit=shared
```

**Comment by JanTvrdik** [comment]
@dg I think this is ready to be merged. Let me know if there is anything that should be improved.

File: tests/Utils/Random.generate().no-mcrypt.phpt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+if (!extension_loaded('mcrypt')) {
+	Tester\Environment::skip('Requires PHP extension mcrypt.');
+}
+
+
+Assert::same( 10, strlen(Random::generate()) );
+Assert::same( 0, strlen(Random::generate(0)) );
+Assert::same( 5, strlen(Random::generate(5)) );
+Assert::same( 200, strlen(Random::generate(200)) );
+
```
Comment (dg) [inline tests/Utils/Random.generate().no-mcrypt.phpt]: Can you add `Assert::random()` test?

**Comment by dg** [comment]
Thanks!

File: tests/Utils/Random.generate().no-mcrypt.phpt (hunk 3 above)
Comment (JanTvrdik) [inline tests/Utils/Random.generate().no-mcrypt.phpt]: Sorry, I don't understand what you mean.

**Pull request merged**